src/lib/Inference.hs

```diff
@@ -1309,9 +1404,35 @@ inferUVar = \case
   UMethodVar v -> do
     MethodBinding _ _ f <- lookupEnv v
     return f
-  UEffectVar _ -> throw NotImplementedErr "inferUVar::UEffectVar"
-  UEffectOpVar _ -> throw NotImplementedErr "inferUVar::UEffectOpVar"
-  UHandlerVar _ -> throw NotImplementedErr "inferUVar::UHandlerVar"
+  UEffectVar _ -> do
+    throw NotImplementedErr "inferUVar::UEffectVar"  -- TODO(alex): implement
+  UEffectOpVar v -> do
+    -- the following should never fail due to operation index since
+    -- users cannot name the "return" operation
+    EffectOpBinding (EffectOpDef effName (OpIdx i)) <- lookupEnv v
+    EffectDef effSourceName _ <- lookupEffectDef effName
+    let holeTy = HandlerDictTy (HandlerDictType effSourceName effName)
+    let holeCon = Con (HandlerHole (AlwaysEqual Nothing) holeTy)
+    Var <$> (emit $ Op (Perform holeCon i))
+  UHandlerVar v -> liftBuilder $ buildHandlerLam v
```

Comment:
Instead of rebuilding the lambdas at every use site, it would be better to construct it once, when you're building the handler, and then to stash it in the `HandlerBinding`

Reply:
Yeah that would be better... we weren't sure up front how large/expensive this would be and adding data fields is... annoying.